### DISCUSSION TITLE
Is it necessary to generate leveldb file from idx file when CommitCompact.

### DISCUSSION THREAD

**Discussion by guol-fnst** [seaweedfs/seaweedfs#3313] [Q&A] [answered]
When deleting objects,  why not  delete the key from leveldb directly.
https://github.com/chrislusf/seaweedfs/blob/957c9a2eb285bfccc7da11efc0d27571c5e90947/weed/storage/needle_map_leveldb.go#L143
 If deleted from leveldb, it seems no need to regerate leveldb from idx when commint compact.
https://github.com/chrislusf/seaweedfs/blob/957c9a2eb285bfccc7da11efc0d27571c5e90947/weed/storage/needle_map_leveldb.go#L75

**Comment by chrislusf** [comment, marked as answer]
`.idx` file is the source of truth.

leveldb file need to be generated from it. 

After compaction, there are a lot of deletions happen. It is simpler to just generate leveldb files again.

**Reply by guol-fnst** [reply]
Yes, I know that logic. 
When deleting objects, a negative needle size is set to the leveldb.
https://github.com/chrislusf/seaweedfs/blob/957c9a2eb285bfccc7da11efc0d27571c5e90947/weed/storage/needle_map_leveldb.go#L143
Why not delete the key from leveldb directly. Leveldb dosen't need to be compacted. It doesn't contain duplicate  needle id

 It is simpler to generate leveldb files, but it cost lots of cpu and disk IO, especially wfs contians billions of small objects files.

**Reply by guol-fnst** [reply]
Got it!

* support read option readDeleted=true
